dbus-tokio/src/connection.rs

```diff
@@ -14,31 +14,44 @@ use tokio_net::driver::Registration;
 /// contact with the D-Bus server.
 pub struct IOResource<C> {
     connection: Arc<C>,
-    registration: Registration,
+    registration: Option<(Registration, std::os::unix::io::RawFd)>,
 }
 
 impl<C: AsRef<Channel> + Process> IOResource<C> {
-    fn poll_internal(&self, ctx: &mut task::Context<'_>) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
+    fn poll_internal(&mut self, ctx: &mut task::Context<'_>) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
         let c: &Channel = (*self.connection).as_ref();
 
         c.read_write(Some(Default::default())).map_err(|_| Error::new_failed("Read/write failed"))?;
         self.connection.process_all();
 
         let w = c.watch();
-        let r = &self.registration;
-        r.register(&mio::unix::EventedFd(&w.fd))?;
-        r.take_read_ready()?;
-        r.take_write_ready()?;
 
-        if w.read { let _ = r.poll_read_ready(ctx)?; };
-        if w.write { let _ = r.poll_write_ready(ctx)?; };
+        match &mut self.registration {
+            None => {
+                let reg = Registration::new(&mio::unix::EventedFd(&w.fd))?;
+                self.registration = Some((reg, w.fd));
+            }
+            Some((_reg, fd)) if *fd == w.fd => (),
+            Some((_reg, _fd)) => unreachable!(),
+        };
+
+        if let Some((r,_)) = &self.registration {
```

Comment:
I suggest to use `expect` (or unwrap) here. Right now the compiler does not get the chance to optimize the branch away.

In my copy of this file at https://github.com/openhab-nodes/wifi-captive-rs/blob/master/src/network_interface/connection.rs#L31 I have set the `registry` field during initialization (see https://github.com/openhab-nodes/wifi-captive-rs/blob/master/src/network_interface/connection.rs#L65), because in the current dbus implementation the file descriptor never changes.

Reply:
Agreed.

Reply:
@davidgraeff Thanks! Your solution looks better:) I have tried copy-paste it, but see this runtime error in test: 

```running 1 test
test connection::method_call ... FAILED

failures:

---- connection::method_call stdout ----
thread 'connection::method_call' panicked at '`spawn_local` called from outside of a task::LocalSet!', src/libcore/option.rs:1190:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.


failures:
    connection::method_call
```